lib/pbench/agent/tool_data_sink.py

```diff
@@ -381,216 +447,354 @@ def tm_log_capture(self):
+            tm["transient_tools"] = transient_tools
+
+            if tm["hostname"] == self.hostname:
+                # The "localhost" tool meister instance does not send data
+                # to the tool data sink, it just writes it locally.
+                tm["posted"] = None
+            elif not transient_tools:
+                # Only Tool Meisters with at least one transient tool will
+                # send data to a data sink, so ignore those Tool Meisters
+                # without any.
+                tm["posted"] = None
+            else:
+                # The `posted` field is "dormant" to start (as set below),
+                # "waiting" when we transition to the "send" state, "dormant"
+                # when we receive data from the target Tool Meister host.
+                tm["posted"] = "dormant"
+
+        # Opportunistically capture the caller's SSH configuration in case
+        # they have host specific configurations they might want to consider
+        # in the future when reviewing historical data.  The host names used
+        # in the config file might be very different from what tools capture
+        # in their output.
+
+        # cp -L  ${HOME}/.ssh/config   ${dir}/ssh.config > /dev/null 2>&1
+        home = os.environ.get("HOME", "")
+        if home:
+            src = str(Path(home) / ".ssh" / "config")
+            dst = str(self.benchmark_run_dir / "ssh.config")
+            try:
+                shutil.copyfile(src, dst)
+            except FileNotFoundError:
+                pass
+            except Exception as exc:
+                self.logger.warning("failed to copy %s to %s: %s", src, dst, exc)
+        # cp -L  /etc/ssh/ssh_config   ${dir}/ > /dev/null 2>&1
+        etc_ssh = Path("/etc") / "ssh"
+        src = str(etc_ssh / "ssh_config")
+        dst = str(self.benchmark_run_dir / "ssh_config")
         try:
-            self.tm_log_capture_thread.join()
-        except Exception:
+            shutil.copyfile(src, dst)
+        except FileNotFoundError:
             pass
-        if self._tm_logging_pubsub:
-            self.logger.debug("unsubscribe tm-logging")
-            self._tm_logging_pubsub.unsubscribe()
-            self.logger.debug("pubsub close tm-logging")
-            self._tm_logging_pubsub.close()
+        except Exception as exc:
+            self.logger.warning("failed to copy %s to %s: %s", src, dst, exc)
+        #
+        # Okay, this is where Python 3 falls down: replacing command line
+        # utilities available via `bash` with Python 3 modules.
+        #
+        # Turns out you can't use `shutil.copytree()` inside a container to
+        # replace `cp -RL`.
+        #
+        # That code attempts to use `os.setxattr()` at the lowest level to
+        # copy all the attributes properly.  But when not running as a real
+        # `root` user in a container, you can't copy all attributes, and a
+        # "Permission denied" exception is raised.
```

Comment:
Arguably this is a bug in `shutil.copytree()`; `cp` doesn't copy xattr unless `--preserve` and even so, `--preserve --no-preserve=xattr` would allow a workaround; but Python's package doesn't provide equivalents except to override `shutil.copyfile2` with `shutil.copyfile` and that doesn't affect the directory handling. Maybe it's worth adding at least some of that detail to the comment for future readers?